adapters/eplanning/eplanning.go

```diff
@@ -14,102 +18,186 @@ import (
-		source, err := verifyImp(&request.Imp[i])
+		imp := request.Imp[i]
+		extImp, err := verifyImp(&imp)
 		if err != nil {
 			errors = append(errors, err)
 			continue
 		}
 
-		// Save valid imp
-		if _, ok := sourceMapper[source]; !ok {
-			sourceMapper[source] = make([]int, 0, totalImps-i)
+		if clientID == "" {
+			clientID = extImp.ClientID
 		}
 
-		sourceMapper[source] = append(sourceMapper[source], i)
+		totalRequests++
+		// Save valid imp
+		name := cleanName(extImp.AdUnitCode)
+		spacesStrings = append(spacesStrings, name+":"+extImp.SizeString)
 	}
 
-	totalRequests := len(sourceMapper)
-
 	if totalRequests == 0 {
 		return nil, errors
 	}
 
-	requests := make([]*adapters.RequestData, 0, totalRequests)
-
 	headers := http.Header{}
 	headers.Add("Content-Type", "application/json")
 	headers.Add("Accept", "application/json")
+	ip := ""
 	if request.Device != nil {
+		ip = request.Device.IP
 		addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
-		addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
+		addHeaderIfNonEmpty(headers, "X-Forwarded-For", ip)
 		addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
 		addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(request.Device.DNT)))
 	}
 
-	imps := make([]openrtb.Imp, len(request.Imp))
-	copy(imps, request.Imp)
+	var pageURL string
+	if request.Site != nil && request.Site.Page != "" {
+		pageURL = request.Site.Page
+	} else {
+		pageURL = defaultPageURL
+	}
 
-	for source, impIds := range sourceMapper {
-		request.Imp = request.Imp[:0]
+	var pageDomain string
+	if request.Site != nil && request.Site.Domain != "" {
+		pageDomain = request.Site.Domain
+	} else {
+		pageDomain = defaultPageURL
+	}
 
-		for i := 0; i < len(impIds); i++ {
-			request.Imp = append(request.Imp, imps[impIds[i]])
-		}
+	uri := adapter.URI + fmt.Sprintf("/%s/%s/%s/%s?r=pbs&ncb=1&ur=%s&e=%s", clientID, dfpClientID, pageDomain, sec, pageURL, strings.Join(spacesStrings, "+"))
 
-		reqJSON, err := json.Marshal(request)
-		if err != nil {
-			errors = append(errors, err)
-			return nil, errors
-		}
+	if request.User != nil && request.User.ID != "" {
+		uri = uri + fmt.Sprintf("&uid=%s", request.User.ID)
+	}
 
-		requestData := adapters.RequestData{
-			Method:  "POST",
-			Uri:     adapter.URI + fmt.Sprintf("/%s", source),
-			Body:    reqJSON,
-			Headers: headers,
-		}
+	if ip != "" {
+		uri = uri + fmt.Sprintf("&ip=%s", ip)
+	}
+
+	var body []byte
+	if adapter.testing {
+		body = []byte("{}")
+	} else {
+		uri = uri + fmt.Sprintf("&rnd=%d", rand.Int())
+		body = nil
+	}
 
-		requests = append(requests, &requestData)
+	requestData := adapters.RequestData{
+		Method:  "GET",
+		Uri:     uri,
+		Body:    body,
+		Headers: headers,
 	}
 
+	requests := []*adapters.RequestData{&requestData}
+
 	return requests, errors
 }
 
-func verifyImp(imp *openrtb.Imp) (string, error) {
+func cleanName(name string) string {
+	for regexp, replacementString := range cleanNameRegexps {
```

Comment:
As @hhhjort said, I'm transitioning away from the project... but this would be a fairly major bug. The iteration order on maps isn't guaranteed, so your three regexes might not run in order.

This looks like a problem, because the 2nd regex might create some `_` chars, while the first one doesn't.

On a related note: can you add a regex to your json-schema to validate inputs more strictly, rather than trying to transform them here? If so, that's a much better approach.

Reply:
If we add a regex to the json-schema, wouldn't nonmatching inputs be rejected? We'd prefer to keep those nonmatching adunits and transform them into what our backend expects, if possible.

Reply:
yes, you're correct. Nonmatching inputs would be rejected.

The way forward is up to you/@hhjort... I can think of a few reasons to strongly recommend JSON schema validation instead.

1. If you change this logic in any way, it would be a breaking API change for publishers. Since PBS host companies will have params saved in their Stored Request backends, this is a pretty serious issue. No promises that the project maintainers would allow it.

2. JSON schemas are returned by the [/bidders/params endpoint](https://prebid.adnxs.com/pbs/v1/bidders/params)... so a UI, for example, can run this validation before anything gets saved in the Stored Request database. Hardcoded validation inside your adapter won't have that benefit.

3. It's consistent with all the other APIs in PBS today, which are "fail loud and fast."

Reply:
So you want to commit to supporting invalid adunit names and have your adapter be responsible for doing the correct thing in converting them to valid names?

Reply:
Yes, we'd prefer the keep the cleanName logic to be consistent with the rest of our systems in which me interact with DFP AdUnits.